### PR TITLE
Task 2.3

### PR DESCRIPTION
**Pull request by DerekAyala**

### CODE REVIEW

File: HW 02/Task 2.3/Lecture.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using System.Xml.Linq;
+
+namespace Task_2._3
```
Comment (IgorKheidorov) [inline HW 02/Task 2.3/Lecture.cs:3]: Fix design, add common base class for training, Lecture, practical lesson (base class (BaseEntity) will include description). For Lecture and practical lesson add one more base class (for example, Lesson) -> 
class  Lecture: Lesson: BaseEntity 
class PracticalLesson: Lesson: BaseEntity.

File: HW 02/Task 2.3/Training.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using System.Xml.Linq;
+
+namespace Task_2._3
+{
+    public class Training
+    {
+        private String textDescription;
```
Comment (IgorKheidorov) [inline HW 02/Task 2.3/Training.cs:5]: private Lesson[] _lessons;

File: HW 02/Task 2.3/Training.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            return true;
+        }
+
+        public object Clone()
+        {
+            Training clonedTraining = new Training();
+            clonedTraining.TextDescription = textDescription;
+            clonedTraining.set = new object[this.set.Length];
+            foreach(var item in set)
+            {
```
Comment (IgorKheidorov) [inline HW 02/Task 2.3/Training.cs:52]: Implement Clone for Lectures and PracticalLesson an simply:
 clonedTraining.Add(item.Clone());

**Review by IgorKheidorov — CHANGES_REQUESTED** [review]
Redesign the architecture

File: HW 02/Task 2.3/Lecture.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        }
+
+        public String Topic
+        {
+            get { return topic; }
+            set { topic = value; }
+        }
+
+        public object Clone()
+        {
```
Comment (IgorKheidorov) [inline HW 02/Task 2.3/Lecture.cs:25]: return  new Lecture(Topic, TextDescription);

File: HW 02/Task 2.3/Lesson.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+namespace Task_2._3
+{
+	public class Lesson
+	{
+        private String textDescription;
+
+        public Lesson(String textDescription)
```
Comment (IgorKheidorov) [inline HW 02/Task 2.3/Lesson.cs]: Fix intensions (Code style)

File: HW 02/Task 2.3/Lesson.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+		{
+			TextDescription = textDescription;
+		}
+
```
Comment (IgorKheidorov) [inline HW 02/Task 2.3/Lesson.cs:13]: Add ICloneable here, to Lesson

File: HW 02/Task 2.3/Training.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        }
+
+        public object Clone()
+        {
+            Training clonedTraining = new Training();
+            clonedTraining.TextDescription = textDescription;
+            clonedTraining._lessons = new Lesson[this._lessons.Length];
+            foreach (var item in _lessons)
+            {
+                if (item is Lecture lecture)
```
Comment (IgorKheidorov) [inline HW 02/Task 2.3/Training.cs]: clonedTraining.Add((Lesson)item.Clone());

If you make Lesson ICloneable (base class for Lecture and PracticalLesson), no need to use if... else  statements

**Review by IgorKheidorov — CHANGES_REQUESTED** [review]
Fix comments, if you as made PR for task 2.3> why I see files for task 2.1 in it? ))))

**Comment by DerekAyala** [comment]
Sorry about files from 2.1, on 2.1 I update git ignore to remove binary files, and I merge main on this brach to get that git ignore

File: HW 02/Task 2.3/PracticalLesson.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
+using System;
+namespace Task_2._3
+{
```
Comment (IgorKheidorov) [inline HW 02/Task 2.3/PracticalLesson.cs:3]: public class PracticalLesson : Lesson. Do not add ICloneable here, it is inherited  from Lesson,

File: HW 02/Task 2.3/Lesson.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+        {
+            TextDescription = textDescription;
+        }
+
+        public string TextDescription
+        {
+            get { return _textDescription; }
+            set { _textDescription = value; }
+        }
+
```
Comment (IgorKheidorov) [inline HW 02/Task 2.3/Lesson.cs:19]: Has to be virtual to be overridden in PracticalLesson and Lecture

File: HW 02/Task 2.3/Training.cs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        }
+
+        public object Clone()
+        {
+            Training clonedTraining = new Training();
+            clonedTraining.TextDescription = textDescription;
+            clonedTraining._lessons = new Lesson[this._lessons.Length];
+            foreach (var item in _lessons)
+            {
+                clonedTraining.Add((Lesson)item.Clone());
```
Comment (IgorKheidorov) [inline HW 02/Task 2.3/Training.cs:53]: Clone from Lesson is called here, not from PracticalLecture and Lecture, thus this is not a deep cloning, you can check this fact ))). Make Clone() in Lesson virtual

**Review by IgorKheidorov — CHANGES_REQUESTED** [review]

**Review by IgorKheidorov — APPROVED** [review]
Now its cool!

**Pull request merged**